extensions/wikia/UserLogin/UserLoginSpecialController.class.php

```diff
@@ -257,72 +257,87 @@ public function getCloseAccountRedirectUrl() {
 	}
 
 	/**
-	 * @brief renders html version that will be inserted into ajax based login interaction
-	 * @details
-	 *   on GET, template partial for an ajax element will render
+	 * Renders html version that will be inserted into ajax based login interaction.
+	 * On GET, template partial for an ajax element will render
 	 */
 	public function dropdown() {
-		$query = $this->app->wg->Request->getValues();
-
-		if ( isset( $query['password'] ) ) {
-			// remove the password from the params to prevent exposiong in into the URL
-			unset($query['password']);
-		}
+		$query = $this->getSanitizedQueryString();
 
-		$this->returnto = $this->getReturnToFromQuery( $query );
-		$this->returntoquery = $this->getReturnToQueryFromQuery( $query );
+		$this->response->setData( [
+			'returnto' => $this->getReturnToFromQuery( $query ),
+			'returntoquery' => $this->getReturnToQueryFromQuery( $query ),
+		] );
 
 		$requestParams = $this->getRequest()->getParams();
 		if ( !empty( $requestParams[ 'registerLink' ] ) ) {
-			$this->registerLink = $requestParams[ 'registerLink' ];
+			$this->response->setVal( 'registerLink',  $requestParams[ 'registerLink' ] );
 		}
+
 		if ( !empty( $requestParams[ 'template' ] ) ) {
 			$this->overrideTemplate( $requestParams[ 'template' ] );
 		}
 	}
 
-	public function getMainPagePartialUrl() {
-		return Title::newMainPage()->getPartialURL();
-	}
-
 	/**
-	 * @param String $title
-	 *
-	 * @return Boolean
+	 * This method returns an array of query parameters that do not contain sensitive information
+	 * This can be used to construct returnto and returntoquery values that are not compromising.
 	 */
-	public function isTitleBlacklisted( $title ) {
-		return AccountNavigationController::isBlacklisted( $title );
+	private function getSanitizedQueryString() {
+		$query = $this->app->wg->Request->getValues();
+		$filterParams = [ 'password', 'username', 'loginToken' ];
+
+		foreach ( $filterParams as $paramName ) {
+			unset( $query[$paramName] );
+		}
+
+		return $query;
 	}
 
-	private function getReturnToFromQuery( $query ) {
-		if( !is_array( $query ) ) {
+	private function getReturnToFromQuery( &$query ) {
+		if ( !is_array( $query ) ) {
 			return '';
 		}
 
-		$returnto = $this->getMainPagePartialUrl();
-		if( isset( $query['title'] ) && !$this->isTitleBlacklisted( $query['title'] ) ) {
-			$returnto = $query['title'] ;
+		// If there's already a returnto here, use it.
+		if ( !empty( $query['returnto'] ) ) {
+			if ( !array_key_exists( 'returntoquery', $query ) ) {
```

Comment:
Well, if I move it to `getReturnToQueryFromQuery` then I have to check the value of `returnto` in that function.  The point is that if I'm using an existing `returnto` value, I don't want to pair it up with whatever random query parameters are on the current URL just because `returntoquery` isn't set.  This code ensures that if we're using an existing `returnto` and `returntoquery` wasn't set, we give it an empty value so that we use that empty value in `getReturnToQueryFromQuery`.